File: mobile/src/api/gstIms.ts
/**
 * GSTN IMS (Invoice Management System) + GSTR-1A API — typed client functions.
 * GAP-101 / board #32. Spec: docs/design/ims-inbox-spec.md
 * Backend: backend/Services/GstService/GstService.Api/Endpoints/GstIms.cs
 *
 * Status vocabulary is the EXACT backend `ImsInvoice.Status` strings:
 *   PENDING | ACCEPTED | REJECTED | PENDING_KEPT
 * Action request values: "ACCEPTED" | "REJECTED" | "PENDING_KEPT".
 * Period format: MMYYYY (e.g. "032026" = March 2026).
 */

import { apiClient } from '../lib/api';

// ─────────────────────────────────────────────────────────────────────────────
// Types — IMS invoices
// ─────────────────────────────────────────────────────────────────────────────

export type ImsInvoiceStatus = 'PENDING' | 'ACCEPTED' | 'REJECTED' | 'PENDING_KEPT';

export type ImsAction = 'ACCEPTED' | 'REJECTED' | 'PENDING_KEPT';

export type ImsSource = 'GSTR-1' | 'IFF';

export interface ImsInvoiceSummary {
  id: string;
  supplierGstin: string;
  supplierName: string;
  invoiceNumber: string;
  /** ISO date YYYY-MM-DD */
  invoiceDate: string;
  invoiceValue: number;
  taxableValue: number;
  igstAmount: number;
  cgstAmount: number;
  sgstAmount: number;
  cessAmount: number;
  /** MMYYYY */
  period: string;
  source: ImsSource | string;
  status: ImsInvoiceStatus;
  deemedAccepted: boolean;
  actionedAt?: string | null;
  actionedBy?: string | null;
}

export interface ImsInvoiceListResponse {
  items: ImsInvoiceSummary[];
  totalCount: number;
  page: number;
  pageSize: number;
}

export interface ImsActionLogEntry {
  id: string;
  action: string;
  previousStatus: string;
  newStatus: string;
  actedAt: string;
  actedBy?: string | null;
  reason?: string | null;
  isBulk: boolean;
}

export interface ImsInvoiceDetail extends ImsInvoiceSummary {
  rejectionReason?: string | null;
  createdAt: string;
  actionLog: ImsActionLogEntry[];
}

export interface ImsSummary {
  period: string;
  pending: number;
  accepted: number;
  rejected: number;
  pendingKept: number;
  total: number;
  /** Count of invoices swept into ACCEPTED by deemed acceptance. */
  deemedAccepted: number;
  /** ISO date — the 14th of the month following the period. */
  gstr2bGenerationDeadline: string;
  gstr2bGenerationPast: boolean;
  totalPendingValue: number;
  totalAcceptedValue: number;
  totalRejectedValue: number;
}

export interface ImsActionResponse {
  invoiceId: string;
  previousStatus: ImsInvoiceStatus;
  newStatus: ImsInvoiceStatus;
  changed: boolean;
  gstnRef?: string | null;
}

export interface ImsBulkActionItem {
  invoiceId: string;
  action: ImsAction;
  reason?: string;
}

export interface ImsBulkInvoiceResult {
  invoiceId: string;
  success: boolean;
  changed: boolean;
  newStatus?: ImsInvoiceStatus | null;
  errorCode?: string | null;
  errorMessage?: string | null;
}

export interface ImsBulkActionResponse {
  totalRequested: number;
  changed: number;
  skipped: number;
  failed: number;
  results: ImsBulkInvoiceResult[];
}

export interface ImsSyncResponse {
  inserted: number;
  skipped: number;
  period: string;
}

// ─────────────────────────────────────────────────────────────────────────────
// Types — GSTR-1A amendments
// ─────────────────────────────────────────────────────────────────────────────

export type Gstr1aStatus = 'DRAFT' | 'SUBMITTED' | 'FILED';

export type Gstr1aAmendmentType =
  | 'B2B_AMENDMENT'
  | 'B2BA'
  | 'CDNR_AMENDMENT'
  | 'CDNRA';

export interface Gstr1aAmendmentSummary {
  id: string;
  originalInvoiceNumber: string;
  originalSupplierGstin: string;
  originalImsInvoiceId?: string | null;
  amendmentType: Gstr1aAmendmentType;
  period: string;
  status: Gstr1aStatus;
  arnNumber?: string | null;
  filedAt?: string | null;
  createdAt: string;
}

export interface Gstr1aListResponse {
  items: Gstr1aAmendmentSummary[];
  totalCount: number;
  page: number;
  pageSize: number;
}

export interface CreateGstr1aAmendmentRequest {
  organizationId: string;
  originalImsInvoiceId?: string | null;
  originalInvoiceNumber: string;
  originalSupplierGstin: string;
  amendmentType: Gstr1aAmendmentType;
  /** Structured corrected figures, serialized client-side. */
  amendmentPayloadJson: string;
  period: string;
}

export interface CreateGstr1aAmendmentResponse {
  amendmentId: string;
  status: Gstr1aStatus;
  period: string;
  amendmentType: Gstr1aAmendmentType;
}

// ─────────────────────────────────────────────────────────────────────────────
// IMS endpoints
// ─────────────────────────────────────────────────────────────────────────────

export async function listImsInvoices(params: {
  organizationId: string;
  period?: string;
  status?: ImsInvoiceStatus;
  supplierGstin?: string;
  search?: string;
  page?: number;
  pageSize?: number;
}): Promise<ImsInvoiceListResponse> {
  const res = await apiClient.get<ImsInvoiceListResponse>('/gst/ims/invoices', {
    params,
  });
  return res.data;
}

export async function getImsInvoice(
  id: string,
  organizationId: string,
): Promise<ImsInvoiceDetail> {
  const res = await apiClient.get<ImsInvoiceDetail>(`/gst/ims/invoices/${id}`, {
    params: { organizationId },
  });
  return res.data;
}

export async function getImsSummary(
  organizationId: string,
  period: string,
): Promise<ImsSummary> {
  const res = await apiClient.get<ImsSummary>('/gst/ims/summary', {
    params: { organizationId, period },
  });
  return res.data;
}

export async function syncImsInvoices(data: {
  organizationId: string;
  gstin: string;
  period: string;
}): Promise<ImsSyncResponse> {
  const res = await apiClient.post<ImsSyncResponse>('/gst/ims/sync', data);
  return res.data;
}

export async function actOnImsInvoice(
  invoiceId: string,
  data: {
    organizationId: string;
    actionedBy: string;
    action: ImsAction;
    reason?: string;
  },
): Promise<ImsActionResponse> {
  const res = await apiClient.post<ImsActionResponse>(
    `/gst/ims/invoices/${invoiceId}/action`,
    data,
  );
  return res.data;
}

export async function bulkActOnImsInvoices(data: {
  organizationId: string;
  actionedBy: string;
  items: ImsBulkActionItem[];
}): Promise<ImsBulkActionResponse> {
  const res = await apiClient.post<ImsBulkActionResponse>(
    '/gst/ims/actions/bulk',
    data,
  );
  return res.data;
}

// ─────────────────────────────────────────────────────────────────────────────
// GSTR-1A endpoints
// ─────────────────────────────────────────────────────────────────────────────

export async function listGstr1aAmendments(params: {
  organizationId: string;
  period?: string;
  status?: Gstr1aStatus;
  page?: number;
  pageSize?: number;
}): Promise<Gstr1aListResponse> {
  const res = await apiClient.get<Gstr1aListResponse>('/gst/gstr1a', { params });
  return res.data;
}

export async function createGstr1aAmendment(
  data: CreateGstr1aAmendmentRequest,
): Promise<CreateGstr1aAmendmentResponse> {
  const res = await apiClient.post<CreateGstr1aAmendmentResponse>(
    '/gst/gstr1a',
    data,
  );
  return res.data;
}

File: backend/Services/FinanceService/Finance.WebApi/Endpoints/Gst/GstIms.cs
using GstService.Application.Gstr1a.Commands.CreateGstr1aAmendment;
using GstService.Application.Gstr1a.Queries.ListGstr1aAmendments;
using GstService.Application.Ims.Commands.ActOnImsInvoice;
using GstService.Application.Ims.Commands.BulkActOnImsInvoices;
using GstService.Application.Ims.Commands.FetchImsInvoices;
using GstService.Application.Ims.Queries.GetImsInvoice;
using GstService.Application.Ims.Queries.GetImsSummary;
using GstService.Application.Ims.Queries.ListImsInvoices;
using MediatR;
using SnapAccount.Shared.Api;

namespace GstService.Api.Endpoints;

/// <summary>
/// IMS (Invoice Management System) endpoints — GAP-101 regulatory requirement.
/// Mandatory from 1 Apr 2026: taxpayers must accept/reject each inward invoice
/// before GSTR-2B is generated. GSTR-3B Table 3 is hard-locked after filing.
///
/// NEW permissions required (see DDL handoff section for seeding):
///   gst.ims.read    — list / get IMS invoices and summary
///   gst.ims.action  — accept / reject / keep-pending IMS invoices
///   gst.ims.sync    — trigger GSTN IMS pull
///   gst.gstr1a.read   — list GSTR-1A amendments
///   gst.gstr1a.create — create GSTR-1A amendment drafts
/// </summary>
public sealed class GstIms : EndpointGroupBase
{
    /// <inheritdoc />
    public override string? GroupName => "/gst";

    /// <inheritdoc />
    public override void Map(RouteGroupBuilder groupBuilder)
    {
        // ── IMS Invoice Inbox ────────────────────────────────────────────────

        // GET /gst/ims/invoices?organizationId=&period=&status=&supplierGstin=&search=&page=&pageSize=
        groupBuilder.MapGet("/ims/invoices", ListImsInvoices)
            .RequireAuthorization().RequireRateLimiting("standard")
            .WithName("ListImsInvoices")
            .WithSummary("List inward invoices visible in the GSTN IMS inbox (paginated, filterable).")
            .WithDescription(
                "Returns supplier-reported inward invoices for a given period. " +
                "Taxpayers must action each invoice before the GSTR-2B generation deadline (14th of following month). " +
                "Rate limit: 100 req/min.");

        // GET /gst/ims/invoices/{id}?organizationId=
        groupBuilder.MapGet("/ims/invoices/{id:guid}", GetImsInvoice)
            .RequireAuthorization().RequireRateLimiting("standard")
            .WithName("GetImsInvoice")
            .WithSummary("Get full detail of a single IMS invoice including action log history.");

        // POST /gst/ims/invoices/{id}/action
        groupBuilder.MapPost("/ims/invoices/{id:guid}/action", ActOnImsInvoice)
            .RequireAuthorization().RequireRateLimiting("gst-write-strict")
            .WithName("ActOnImsInvoice")
            .WithSummary("Accept, reject, or mark as pending-kept a single IMS invoice. Idempotent.")
            .WithDescription(
                "Action is idempotent for the same status transition. " +
                "Cannot reject an accepted invoice — use POST /gst/gstr1a instead. " +
                "Rate limit: 30 req/min (GSTN API cost).");

        // POST /gst/ims/actions/bulk
        groupBuilder.MapPost("/ims/actions/bulk", BulkActOnImsInvoices)
            .RequireAuthorization().RequireRateLimiting("gst-write-strict")
            .WithName("BulkActOnImsInvoices")
            .WithSummary("Apply IMS actions to up to 100 invoices in a single request.")
            .WithDescription(
                "Bulk limit: 100 invoices per request (GSTN rate limit). " +
                "Returns per-invoice results including success/failure/skipped counts. " +
                "Rate limit: 30 req/min.");

        // GET /gst/ims/summary?organizationId=&period=
        groupBuilder.MapGet("/ims/summary", GetImsSummary)
            .RequireAuthorization().RequireRateLimiting("standard")
            .WithName("GetImsSummary")
            .WithSummary("IMS status counts by period + GSTR-2B generation deadline information.")
            .WithDescription(
                "Returns PENDING/ACCEPTED/REJECTED/PENDING_KEPT counts, total invoice values, " +
                "GSTR-2B deadline date (14th of following month), and whether deemed-acceptance has triggered.");

        // POST /gst/ims/sync
        groupBuilder.MapPost("/ims/sync", SyncImsInvoices)
            .RequireAuthorization().RequireRateLimiting("gst-write-strict")
            .WithName("SyncImsInvoices")
            .WithSummary("Pull inward invoices from GSTN IMS into local store for the given period.")
            .WithDescription(
                "Upsert semantics: new invoices inserted; existing records (matched by supplier GSTIN + invoice number + period) " +
                "are left unchanged to preserve local action status. " +
                "Safe to call repeatedly (idempotent). Rate limit: 30 req/min.");

        // ── GSTR-1A Amendments ───────────────────────────────────────────────

        // POST /gst/gstr1a
        groupBuilder.MapPost("/gstr1a", CreateGstr1aAmendment)
            .RequireAuthorization().RequireRateLimiting("gst-write-strict")
            .WithName("CreateGstr1aAmendment")
            .WithSummary("Create a GSTR-1A amendment draft (only mechanism to correct GSTR-3B Table 3 post-filing).")
            .WithDescription(
                "GSTR-1A is the correction route when a taxpayer has rejected an IMS invoice and the supplier needs to amend their return. " +
                "Amendment types: B2B_AMENDMENT, B2BA, CDNR_AMENDMENT, CDNRA. " +
                "Rate limit: 30 req/min.");

        // GET /gst/gstr1a?organizationId=&period=&status=&page=&pageSize=
        groupBuilder.MapGet("/gstr1a", ListGstr1aAmendments)
            .RequireAuthorization().RequireRateLimiting("standard")
            .WithName("ListGstr1aAmendments")
            .WithSummary("List GSTR-1A amendment drafts for an organisation.");
    }

    // ── Handlers ─────────────────────────────────────────────────────────────

    private static async Task<IResult> ListImsInvoices(
        ISender sender,
        Guid? organizationId = null,
        string? period = null,
        string? status = null,
        string? supplierGstin = null,
        string? search = null,
        int page = 1,
        int pageSize = 20)
    {
        if (!organizationId.HasValue)
            return Results.BadRequest(new { error = "organizationId query parameter is required.", code = "GST.MissingOrganizationId" });

        var result = await sender.Send(new ListImsInvoicesQuery(
            organizationId.Value, period, status, supplierGstin, search, page, pageSize));
        return result.IsSuccess ? Results.Ok(result.Value) : result.Error.ToHttpResult();
    }

    private static async Task<IResult> GetImsInvoice(Guid id, ISender sender, Guid? organizationId = null)
    {
        if (!organizationId.HasValue)
            return Results.BadRequest(new { error = "organizationId query parameter is required.", code = "GST.MissingOrganizationId" });

        var result = await sender.Send(new GetImsInvoiceQuery(id, organizationId.Value));
        return result.IsSuccess ? Results.Ok(result.Value) : result.Error.ToHttpResult();
    }

    private static async Task<IResult> ActOnImsInvoice(
        Guid id, ActOnImsInvoiceRequest req, ISender sender)
    {
        var result = await sender.Send(new ActOnImsInvoiceCommand(
            InvoiceId: id,
            OrganizationId: req.OrganizationId,
            Action: req.Action,
            Reason: req.Reason,
            ActionedBy: req.ActionedBy));
        return result.IsSuccess ? Results.Ok(result.Value) : result.Error.ToHttpResult();
    }

    private static async Task<IResult> BulkActOnImsInvoices(BulkActOnImsInvoicesRequest req, ISender sender)
    {
        var items = req.Items
            .Select(i => new BulkImsActionItem(i.InvoiceId, i.Action, i.Reason))
            .ToList();
        var result = await sender.Send(new BulkActOnImsInvoicesCommand(req.OrganizationId, req.ActionedBy, items));
        return result.IsSuccess ? Results.Ok(result.Value) : result.Error.ToHttpResult();
    }

    private static async Task<IResult> GetImsSummary(
        ISender sender, Guid? organizationId = null, string? period = null)
    {
        if (!organizationId.HasValue)
            return Results.BadRequest(new { error = "organizationId query parameter is required.", code = "GST.MissingOrganizationId" });
        if (string.IsNullOrEmpty(period))
            return Results.BadRequest(new { error = "period query parameter is required (MMYYYY format).", code = "GST.MissingPeriod" });

        var result = await sender.Send(new GetImsSummaryQuery(organizationId.Value, period));
        return result.IsSuccess ? Results.Ok(result.Value) : result.Error.ToHttpResult();
    }

    private static async Task<IResult> SyncImsInvoices(SyncImsInvoicesRequest req, ISender sender)
    {
        var result = await sender.Send(new FetchImsInvoicesCommand(req.OrganizationId, req.Gstin, req.Period));
        return result.IsSuccess ? Results.Ok(result.Value) : result.Error.ToHttpResult();
    }

    private static async Task<IResult> CreateGstr1aAmendment(CreateGstr1aAmendmentRequest req, ISender sender)
    {
        var result = await sender.Send(new CreateGstr1aAmendmentCommand(
            OrganizationId: req.OrganizationId,
            OriginalImsInvoiceId: req.OriginalImsInvoiceId,
            OriginalInvoiceNumber: req.OriginalInvoiceNumber,
            OriginalSupplierGstin: req.OriginalSupplierGstin,
            AmendmentType: req.AmendmentType,
            AmendmentPayloadJson: req.AmendmentPayloadJson,
            Period: req.Period));
        return result.IsSuccess
            ? Results.Created($"/gst/gstr1a/{result.Value.AmendmentId}", result.Value)
            : result.Error.ToHttpResult();
    }

    private static async Task<IResult> ListGstr1aAmendments(
        ISender sender,
        Guid? organizationId = null,
        string? period = null,
        string? status = null,
        int page = 1,
        int pageSize = 20)
    {
        if (!organizationId.HasValue)
            return Results.BadRequest(new { error = "organizationId query parameter is required.", code = "GST.MissingOrganizationId" });

        var result = await sender.Send(new ListGstr1aAmendmentsQuery(organizationId.Value, period, status, page, pageSize));
        return result.IsSuccess ? Results.Ok(result.Value) : result.Error.ToHttpResult();
    }
}

// ── Request DTOs ──────────────────────────────────────────────────────────────

/// <summary>Request body for acting on a single IMS invoice.</summary>
internal record ActOnImsInvoiceRequest(
    Guid OrganizationId,
    Guid ActionedBy,
    string Action,
    string? Reason = null);

/// <summary>Request body for bulk IMS invoice actions.</summary>
internal record BulkActOnImsInvoicesRequest(
    Guid OrganizationId,
    Guid ActionedBy,
    IReadOnlyList<BulkImsActionItemRequest> Items);

/// <summary>One item in a bulk action request body.</summary>
internal record BulkImsActionItemRequest(
    Guid InvoiceId,
    string Action,
    string? Reason = null);

/// <summary>Request body for GSTN IMS sync.</summary>
internal record SyncImsInvoicesRequest(
    Guid OrganizationId,
    string Gstin,
    string Period);

/// <summary>Request body for creating a GSTR-1A amendment draft.</summary>
internal record CreateGstr1aAmendmentRequest(
    Guid OrganizationId,
    string OriginalInvoiceNumber,
    string OriginalSupplierGstin,
    string AmendmentType,
    string AmendmentPayloadJson,
    string Period,
    Guid? OriginalImsInvoiceId = null);
